### DISCUSSION TITLE
Error 400 when upgrading to websocket transport

### DISCUSSION THREAD

**Discussion by DaRealCodeWritten** [miguelgrinberg/Flask-SocketIO#1864] [Q&A]
**Describe the bug**
Flask-SocketsIO + Flask, attempts to use websockets as transport results in a 400 response code and a traceback on the client side, using `simple-websocket` as the WS framework and NGINX as the reverse proxy
Nginx config is as follows

```
server {
    server_name czvr-bot.xyz;
    location / {
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_set_header X-Real-IP $remote_addr;
        proxy_pass http://czvr-bot.xyz;
    }

    # SSL (if wanted)
    listen 443 ssl;
    ssl_certificate fullchain1.pem;
    ssl_certificate_key privkey1.pem;
}
```

**Expected behavior**
Website upgrades to WSS for the connection between the site and the bot

**Actual behavior**
Client throws the below traceback and the website refuses the upgrade

**Logs**
Python-SocketIO client (A discord bot)
```
Traceback (most recent call last):
  File "C:\Users\Eric Lysne\PycharmProjects\VancouverSite\bot.py", line 318, in inner
    await sock.connect("https://czvr-bot.xyz", transports="websocket")
  File "C:\Users\Eric Lysne\PycharmProjects\VancouverSite\venv\lib\site-packages\socketio\asyncio_client.py", line 149, in connect
    raise exceptions.ConnectionError(exc.args[0]) from None
socketio.exceptions.ConnectionError: Connection error
```
Flask-SocketIO server (A flask app)
```
Invalid websocket upgrade
INFO:engineio.server:Invalid websocket upgrade
INFO:werkzeug:127.0.0.1 - - [08/Aug/2022 01:16:37] "GET /socket.io/?transport=websocket&EIO=4&t=1659946597.8447661 HTTP/1.0" 400 -
```

**What i've tried so far**
Changing the `https://` in the  `sock.connect()` call to `wss://`
Modifying the order of the proxy directives in NGINX
Adding HTTP version directive in NGINX

**Comment by miguelgrinberg** [comment]
I converted this to a discussion because at this time I don't really have an indication that there is a bug here.

The "Invalid websocket upgrade" error occurs when the client sends a malformed websocket connection request, such as for example, one that does not include the proper upgrade HTTP headers. My guess is that your nginx configuration is causing this, and not your client. I suggest you look at the example configuration in the Flask-SocketIO documentation and try that approach to see if it helps.

**Reply by DaRealCodeWritten** [reply]
So i tried adapting the sample NGINX config to what works for me, still no dice, does the URL the client connects to need to be prefixed with wss:// rather than https:// or does it matter?
New NGINX config:
```
server {
    server_name czvr-bot.xyz;
    location / {
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_set_header X-Real-IP $remote_addr;
        proxy_pass http://czvr-bot.xyz;
    }
    location /socket.io {
        proxy_http_version 1.1;
        proxy_buffering off;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_pass http://czvr-bot.xyz/socket.io;
    }
    # SSL (if wanted)
    listen 443 ssl;
    ssl_certificate fullchain1.pem;
    ssl_certificate_key privkey1.pem;
}
```
Relevant code snips:
Bot
```
sock = socketio.AsyncClient(logger=True, engineio_logger=True)
...
def sockio():
    async def inner():  # Inner? I hardly know'er!
        await sock.connect("https://czvr-bot.xyz", transports="websocket")
        await sock.wait()
    bot.loop.create_task(inner())
```
Website
```
app = Flask(__name__)
app.secret_key = app_config["FLASK_CLIENT_SECRET"]
sock = SocketIO(app, logger=True, engineio_logger=True)
```

**Reply by miguelgrinberg** [reply]
The `https` vs `wss` prefix does not matter.

Here are some things that you can try:

- Remove SSL. Does it work then? My guess is that SSL is not part of the problem, so I don't expect it to work either.
- Remove Nginx. Does it work then? I expect this will make things work.

Your Nginx configuration appears to be recursive, you are forwarding your incoming requests to the same server given in the server name directive. This looks wrong to me, normally Nginx forwards to a local server given as `localhost:port`.

**Reply by DaRealCodeWritten** [reply]
The server proxy is not recursive, czvr-bot.xyz is defined as an `upstream` which directs to `127.0.0.1:6880`
What i'm noticing is that when the bot calls upon the /socket.io endpoint, its using HTTP version 1.0, despite the NGINX proxy directive telling it to use 1.1

Side note: I cannot remove NGINX as it handles SSL, removing it would make connections to the webserver impossible

**Reply by miguelgrinberg** [reply]
I suggested you remove SSL as well, go back and read my suggestions again. Preempting your next response that your bot platform can only use SSL connections, my suggestion is that you create a simple Socket.IO client using the python-socketio package and debug your connections with it.

**Reply by DaRealCodeWritten** [reply]
True, I'll take nginx out of the equation for now and give 'er a try

**Reply by DaRealCodeWritten** [reply]
Update: Removed NGINX and connected direct to the webserver, it accepted the upgrade and is functioning as expected, im not sure what NGINX is doing to prevent the upgrade

**Reply by DaRealCodeWritten** [reply]
Another update on the situation, i added
```
map $http_upgrade $connection_upgrade {
    default upgrade;
    '' close;
}
```
to NGINX and reattempted connecting through NGINX, and the webserver accepted the upgrade to WSS as expected

**Reply by miguelgrinberg** [reply]
Not sure why this makes a difference for you, as I have never needed to include this. But in any case, it works so maybe the why isn't that important. :)

**Reply by DaRealCodeWritten** [reply]
Might be because i removed the `include proxy_params;` since nginx whines about it if its included, which might be because im using nginx for windows
